src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ExhaustivenessComputer.java

```diff
@@ -673,5 +772,434 @@ public String toString() {
+                                 .collect(Collectors.toCollection(HashSet::new));
+
+                //remove the permitted subtypes that are not needed to achieve exhaustivity
+                boolean reduced = false;
+
+                for (Iterator<BindingPattern> it = viablePermittedPatterns.iterator(); it.hasNext(); ) {
+                    BindingPattern current = it.next();
+                    Set<BindingPattern> reducedPermittedPatterns = new HashSet<>(viablePermittedPatterns);
+
+                    reducedPermittedPatterns.remove(current);
+
+                    Set<PatternDescription> replaced =
+                            replace(inMissingPatterns, toExpand, reducedPermittedPatterns);
+
+                    if (computeCoverage(selectorType, joinSets(basePatterns, replaced), true).covered()) {
+                        it.remove();
+                        reduced = true;
+                    }
+                }
+
+                if (!reduced) {
+                    //if all immediate permitted subtypes are needed
+                    //give up, and simply use the current pattern:
+                    return inMissingPatterns;
+                }
+
+                Set<PatternDescription> currentMissingPatterns =
+                        replace(inMissingPatterns, toExpand, viablePermittedPatterns);
+
+                //try to recursively expand on each viable pattern:
+                for (PatternDescription viable : viablePermittedPatterns) {
+                    currentMissingPatterns = expandMissingPatternDescriptions(selectorType, targetType,
+                                                                              viable, basePatterns,
+                                                                              currentMissingPatterns);
+                }
+
+                return currentMissingPatterns;
+            } else if ((bp.type.tsym.flags_field & Flags.RECORD) != 0 &&
+                       //only expand record types into record patterns if there's a chance it may change the outcome
+                       //i.e. there is a record pattern in at the spot in the original base patterns:
+                       hasMatchingRecordPattern(basePatterns, inMissingPatterns, toExpand)) {
+                //if there is a binding pattern at a place where the original based patterns
+                //have a record pattern, try to expand the binding pattern into a record pattern
+                //create all possible combinations of record pattern components:
+                Type[] componentTypes = ((ClassSymbol) bp.type.tsym).getRecordComponents()
+                        .map(r -> types.memberType(bp.type, r))
+                        .toArray(s -> new Type[s]);
+                List<List<Type>> combinatorialNestedTypes = List.of(List.nil());
+
+                for (Type componentType : componentTypes) {
+                    List<Type> variants;
```

Comment:
Is this the same conceptually to `viablePermittedPatterns` for the binding patterns case? If yes, we can reuse the same name (just to match the concepts).